### PR TITLE
Improve clarity of CheckedExceptionNotThrown description

### PR DESCRIPTION
**Pull request by delanym**
Improve clarity of CheckedExceptionNotThrown description re: #3774

### CODE REVIEW

File: core/src/main/java/com/google/errorprone/bugpatterns/CheckedExceptionNotThrown.java [hunk 1]
```diff
@@ -64,8 +64,8 @@
 /** Flags checked exceptions which are claimed to be thrown, but are not. */
 @BugPattern(
     summary =
-        "This method cannot throw a checked exception that it claims to. This may cause consumers"
-            + " of the API to incorrectly attempt to handle, or propagate, this exception.",
+        "This method is overridable and cannot throw any of the checked exceptions it claims to. This may cause consumers"
```
Comment (tbroyer) [inline core/src/main/java/com/google/errorprone/bugpatterns/CheckedExceptionNotThrown.java:67]: Shouldn't that say "final" rather than "overridable"?

Also, about the "any of the checked exceptions", is that really the current behavior? Shouldn't that be "one of the checked exceptions"? (in which case, the original phrasing was clearer; what I find unclear, as a non-native English speaker, is the "cannot", "will not" or "does not", or "never throws" would be clearer)

File: docs/bugpattern/CheckedExceptionNotThrown.md [hunk 2]
```diff
@@ -18,6 +16,8 @@ Response handle(Request request) {
...
 ```
 
-Including unthrown exceptions can be reasonable where the method is overridable,
-as overriding methods will not be able to declare that they throw any exceptions
-not included.
+Unfortunately this can lead to call sites being forced to explicitly handle or
+propagate exceptions which provably can never occur. It may also lead readers of
+the code to search for possibly throwing paths where none exist.
+
+The rule checks only whether the method is overridable, not whether it is overridden.
```
Comment (tbroyer) [inline docs/bugpattern/CheckedExceptionNotThrown.md:23]: It would probably be enough, IMO, to add to the previous wording: "; this check will thus only flag non-overridable methods." (or "not flag overridable methods")

**Review by tbroyer — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/google/errorprone/bugpatterns/CheckedExceptionNotThrown.java (hunk 1 above)
Comment (delanym) [reply core/src/main/java/com/google/errorprone/bugpatterns/CheckedExceptionNotThrown.java:67]: Actually it should say "non-overridable". Final methods are non-overridable AND they also can't be hidden. Static methods are non-overridable but can be hidden. Private makes the method un-overridable through being inaccessible. I confirmed the check considers all these cases.

Yes you're right the original is clearer, but even clearer is to avoid the indefinite article, so "one or more".

Even as native English speaker we often miss points like this. To say "cannot" might suggest the method is trying to "throw new IOException" or the like. But this is about how the method throws/propagates unhandled exceptions. The method "will never" be required to throw the exception. "does not" is something to consider at runtime.

File: docs/bugpattern/CheckedExceptionNotThrown.md (hunk 2 above)
Comment (delanym) [reply docs/bugpattern/CheckedExceptionNotThrown.md:23]: I wanted to answer the question why Java allows redundant throws in the opening statement and not leave it for later. Currently the two statements stand in isolation, yet this is the crux of how the issue comes up in the first place.

As I look at the example now I see it showing the undesirable propagation but not the main issue: checkArgument doesn't throw IOException. I'll try come up with a better example

Comment (delanym) [reply docs/bugpattern/CheckedExceptionNotThrown.md:23]: @tbroyer do you understand this "It may also lead readers of the code to search for possibly throwing paths where none exist." ?

File: docs/bugpattern/CheckedExceptionNotThrown.md [hunk 3]
```diff
@@ -1,11 +1,11 @@
...
-don't. This can lead to call sites being forced to explicitly handle or
+Java allows methods to declare that they throw checked exceptions even when they don’t.
+This can lead to call sites being forced to explicitly handle or
 propagate exceptions which provably can never occur. It may also lead readers of
 the code to search for possibly throwing paths where none exist.
 
 ```java
 private static void validateRequest(Request request) throws IOException {
-  checkArgument(request.hasFoo(), "foo must be specified");
+  System.out.println("Request " + request.name + " validated");
```
Comment (tbroyer) [inline docs/bugpattern/CheckedExceptionNotThrown.md:8]: That doesn't really look like a "validateRequest" implementation. Maybe replace checkArgument with the equivalent non-Guava code instead if you think it's unclear?
```suggestion
  if (!request.hasFoo()) {
    throw new IllegalArgumentException("foo must be specified");
  }
```

File: docs/bugpattern/CheckedExceptionNotThrown.md [hunk 4]
```diff
@@ -1,11 +1,11 @@
-Java allows methods to declare that they throw checked exceptions even when they
-don't. This can lead to call sites being forced to explicitly handle or
+Java allows methods to declare that they throw checked exceptions even when they don’t.
+This can lead to call sites being forced to explicitly handle or
```
Comment (tbroyer) [inline docs/bugpattern/CheckedExceptionNotThrown.md:2]: nit: unnecessary change

File: docs/bugpattern/CheckedExceptionNotThrown.md [hunk 5]
```diff
@@ -18,6 +18,4 @@ Response handle(Request request) {
 }
 ```
 
-Including unthrown exceptions can be reasonable where the method is overridable,
-as overriding methods will not be able to declare that they throw any exceptions
-not included.
+The check will not flag overridable methods.
```
Comment (tbroyer) [inline docs/bugpattern/CheckedExceptionNotThrown.md:21]: I think the previous text was clearer in that it described **why** that would be the case ("it can be reasonable"). I'd rather **add** to the previous text than replace it.
```suggestion
Including unthrown exceptions can be reasonable where the method is overridable,
as overriding methods will not be able to declare that they throw any exceptions
not included. The check will therefore not flag overridable methods.
```

Comment (delanym) [reply docs/bugpattern/CheckedExceptionNotThrown.md:24]: Moved up and parenthesized. The last sentence describes the scope of the check, as a kind of disclaimer. Once the check is adapted to inspect overridable methods it might say "The check will flag non-overridable methods and methods that declare throwing a checked exception where none of the overridden methods throw that exception."
